src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -259,417 +258,482 @@ AND NOT EXISTS (
-                            }
-                            else {
-                                positives.add(attrValue);
-                            }
-                        }
-
-                        if (!positives.isEmpty()) {
-                            criteria.add(this.addAttributeFilterSubquery(attrib, positives));
-                        }
-
-                        if (!negatives.isEmpty()) {
-                            criteria.add(Restrictions.not(
-                                this.addAttributeFilterSubquery(attrib, negatives)));
-                        }
+                List<Predicate> inclusionPredicates = new ArrayList<>();
+                List<Predicate> exclusionPredicates = new ArrayList<>();
+                for (String attributeValue : attributeFilters) {
+                    String sanitized = attributeValue.startsWith("!") ?
+                        sanitizeMatchesFilter(attributeValue.substring(1)) :
+                        sanitizeMatchesFilter(attributeValue);
+
+                    if (attributeValue.startsWith("!")) {
+                        exclusionPredicates.add(buildPoolAttributePredicate(builder, query, root, key,
+                            attr -> ilike(builder, attr, sanitized)));
                     }
                     else {
-                        criteria.add(this.addAttributeFilterSubquery(attrib, attributeFilters));
+                        inclusionPredicates.add(buildPoolAttributePredicate(builder, query, root, key,
+                            attr -> ilike(builder, attr, sanitized)));
                     }
                 }
+
+                if (!inclusionPredicates.isEmpty()) {
+                    predicates.add(builder.or(inclusionPredicates.toArray(new Predicate[0])));
+                }
+
+                if (!exclusionPredicates.isEmpty()) {
+                    predicates.add(builder.or(exclusionPredicates.toArray(new Predicate[0])).not());
+                }
             }
         }
 
-        // Impl note:
-        // Hibernate has an issue with properly hydrating objects within collections of the pool
-        // when only a subset of the collection matches the criteria. To work around this, we pull
-        // the ID list from the main filtering query, then pull the pools again using the ID list.
-        // This also makes it easier to eventually start using a cursor, since the distinct entity
-        // functionality doesn't work with cursors.
-
-        List<String> poolIds = criteria.list();
+        return predicates;
+    }
 
-        if (poolIds != null && !poolIds.isEmpty()) {
-            criteria = this.currentSession()
-                .createCriteria(Pool.class)
-                .createAlias("product", "Product");
-            criteria.add(CPRestrictions.in("id", poolIds));
+    private Predicate getConsumerPredicate(CriteriaQuery<?> query, Root<Pool> root, CriteriaBuilder builder,
```

Comment:
Something we can do to cleanup all of these and make the base code use a bit more of a functional style (as in the programming paradigm), is returning an `Optional<Predicate>` so we can do something like this:

```
this.getConsumerPredicate(...)
  .ifPresent(predicates::add);
```

Then we also don't need to worry about nulls hitting the predicate collection, which, honestly, I have no idea how well Hibernate handles it.

Reply:
I like that! Much easier to read. Updated throughout